contributors/design-proposals/scheduling/scheduling-policy.md

```diff
@@ -0,0 +1,302 @@
+
+```yaml
+apiVersion: policy/v1alpha1
+kind: SchedulingPolicy
+metadata:
+  name: mySchedulingPolicy
+spec:
+  action: allow    
+  rules:
+    namespaces:
+      - key: team
+        operator: Exists
+    podSelector: {}
+    schedulerNames:
+       match: ["green-scheduler","my-scheduler"]
+```
+
+An empty list of schedulerNames will allow usage of all schedulers:
+
+```yaml
+apiVersion: policy/v1alpha1
+kind: SchedulingPolicy
+metadata:
+  name: mySchedulingPolicy
+spec:
+  action: allow
+  rules:
+    namespaces:
+      - key: team
+        operator: Exists
+    podSelector: {}    
+    schedulerNames:
+       match: []
+```
+
+note: this policy has no effect as we allow all when no policy is set.
+
+### Tolerations
+
+Toleration usage can be regulated using fine-grain rules with `tolerations` field. If specifying multiple `tolerations`, pod will be scheduled if one of the tolerations is satisfied.
+
+
+#### Allowed
+
+This allows pods with tolerations the following:
+- tolerations that tolerates taints with key named `mykey` that has a value `value` and with a `NoSchedule` effect.
+- tolerations that tolerates taints with key `other_key` that has a `NoExecute` effect.
+
+```yaml
+apiVersion: policy/v1alpha1
+kind: SchedulingPolicy
+metadata:
+  name: mySchedulingPolicy
+spec:
+  action: allow   
+  rules:
+    namespaces:
+      - key: team
+        operator: Exists
+    podSelector: {}    
+    tolerations:
+       match:
+        - key: "mykey"
+          operator: "Equal"
+          value: "value"
+          effects: "NoSchedule"
+        - key: "other_key"
+          operator: "Exists"
+          effect: "NoExecute"
+```
+
+
+### Priority classes
+
+Priority class usage can be regulated using fine-grain rules with `priorityClasseName` field.
+
+##### Allow
+
+In this example, we only allow the `critical-job` priority.
+
+
+```yaml
+apiVersion: policy/v1alpha1
+kind: SchedulingPolicy
+metadata:
+  name: mySchedulingPolicy
+spec:
+  action: allow
+  rules:
+    namespaces:
+      - key: team
+        operator: Exists
+    podSelector: {}
+    priorityClasseNames:
+      - match: "critical-priority"
+```
+
+
+### Node Selector
+
+The `nodeSelector` field makes it possible to specify what pods are allowed or denied based on their nodeSelectors.
+
+#### Examples
+
+##### Complete policy
+
+```yaml
+apiVersion: policy/v1alpha1
+kind: SchedulingPolicy
+metadata:
+  name: mySchedulingPolicy
+spec:
+  action: deny   
+  rules:
+    namespaces:
+      - key: team
+        operator: Exists
+    podSelector: {}
+    nodeSelectors:
+      match:
+        - failure-domain.beta.kubernetes.io/region: ""
+        - disk: "ssd"
+        - disk: "premium"
+```
+
+
+In this example, pods cannot be scheduled if they have all of the following at the same time:
+-  `disk: ssd` nodeSelector
+-  `disk: hdd` nodeSelector
+-  `failure-domain.beta.kubernetes.io/region` nodeSelector with any value.
```

Comment:
What if they have these 3, plus one other? Or only 2? See above about matching maps.

Reply:
It depends on the algorithm, as we said we should match everything to take the action ? cc @bsalamat

Reply:
I agree with match all for a given rule. One could create multiple policies to cover "OR" cases. For example, you want to deny any of nodeSelector "disk: ssd" or nodeSelector "disk:premimum", you could create two policies with deny action. One has
```
nodeSelectors:
  match:
    - disk: "ssd"
```
and the other:
```
nodeSelectors:
  match:
    - disk: "premium"
```